### DISCUSSION TITLE
Guidance on writing cross platform library code with C# 8.0

### DISCUSSION THREAD

**Discussion by YairHalberstadt** [dotnet/csharplang#2002] [Q&A]
In this recent blog post it was confirmed that various new C# 8.0 features would not be coming to the .Net Framework.

https://blogs.msdn.microsoft.com/dotnet/2018/11/12/building-c-8-0/

I'm sure the C# team has put some effort into writing library code with these new features, and has considered whether such features would bifurcate the ecosystem.

As such is there any, or will there be any, guidance available as to using the new features in cross platform libraries? Should they just not be used at all? Are there any ways around this problem, such as using conditional compilation for the default interface methods or their bodies?

etc.

**Comment by svick** [comment]
I *think* language features that will only work on some frameworks will behave almost the same as APIs that are only available on some frameworks.

This means that:

* If you're writing a pure .Net Standard library, you won't be able to use those features (for now?).
* If you're writing a pure .Net Core 3.0 library (which should be rare), feel free to use those features.
* If you're writing a multi-targeted library, you will be able to use the same conditional compilation approaches that you can use now. Though I think that won't make much sense for DIM.

I do think it would mean that these features will have very limited use in libraries for a while.

**Comment by DavidArno** [comment]
@svick,
I'm not sure you are right with your first point. If you are targeting .Net Standard 2.1, then all of the features will be available (I think).

**Comment by YairHalberstadt** [comment]
I could imagine using conditional compilation to add a DIM in .Net Standard 2.1, or an extension method in .Net Framework, though I'm not sure how practical that would be.

Similarly, I could imagine declaring a body for an interface member in .Net Standard 2.1, but leave it unimplemented in .Net Framework and force the consumer to implement it.

**Comment by svick** [comment]
@DavidArno According to [the proposal](https://github.com/dotnet/csharplang/blob/master/proposals/default-interface-methods.md#clr-support-api), DIM will require a `RuntimeFeature` constant to exist. [That constant is defined in .Net Core 3.0](https://github.com/dotnet/corefx/blob/dcc1bef/src/System.Runtime/ref/System.Runtime.cs#L6756) (behind an `#if`), but [it's not currently defined in .Net Standard 2.1](https://github.com/dotnet/standard/blob/91af1df/netstandard/ref/System.Runtime.CompilerServices.cs#L499-L503). I think that's not going to change, because I doubt DIM will work on all active .Net Standard platforms (UWP, Xamarin, Unity) in the short term.

**Comment by svick** [comment]
@YairHalberstadt 

> I could imagine using conditional compilation to add a DIM in .Net Standard 2.1, or an extension method in .Net Framework, though I'm not sure how practical that would be.

I don't think that would work well. Especially since it would require you to write more code than just using an extension method everywhere.

>  I could imagine declaring a body for an interface member in .Net Standard 2.1, but leave it unimplemented in .Net Framework and force the consumer to implement it.

That actually makes sense. Those on platforms that don't support DIM would only lose some convenience and the amount of code you would have write twice probably wouldn't be that bad.

**Comment by DavidArno** [comment]
@svick, that's useful to know. Thanks.

**Comment by jnm2** [comment]
> I could imagine declaring a body for an interface member in .Net Standard 2.1, but leave it unimplemented in .Net Framework and force the consumer to implement it.

This works if the interface has never existed before, but it fails in DIM's primary mission of enabling evolution after that point without breaking existing implementors.

**Comment by YairHalberstadt** [comment]
https://github.com/dotnet/standard/pull/1019
